kevm_pyk/src/kevm_pyk/solc_to_k.py

```diff
@@ -37,6 +39,198 @@
+
+    def __init__(self, contract_name: str, contract_json: Dict, foundry: bool = False) -> None:
+
+        def _get_method_abi(_mname: str) -> Dict:
+            for _method in contract_json['abi']:
+                if _method['type'] == 'function' and _method['name'] == _mname:
+                    return _method
+            raise ValueError(f'Method not found in abi: {_mname}')
+
+        self.name = contract_name
+        self.storage = contract_json['storageLayout']
+        self.bytecode = (contract_json['evm']['deployedBytecode']['object'] if not foundry else contract_json['deployedBytecode']['object'])
+        method_identifiers = contract_json['evm']['methodIdentifiers'] if not foundry else contract_json['methodIdentifiers']
+        self.methods = []
+        for msig in method_identifiers:
+            mname = msig.split('(')[0]
+            mid = int(method_identifiers[msig], 16)
+            self.methods.append(Contract.Method(mname, mid, _get_method_abi(mname), contract_name, self.sort_method))
+        self.fields = []
+        for storage in contract_json['storageLayout']['storage']:
+            if storage['offset'] != 0:
+                raise ValueError(f'Unsupported nonzero offset for contract {self.name} storage slot: {storage["label"]}')
+            self.fields.append(Contract.Field(storage['label'], int(storage['slot']), storage['type'], self.name, self.sort_field, contract_json['storageLayout']['types']))
+
+    @property
+    def sort(self) -> KSort:
+        return KSort(f'{self.name}Contract')
+
+    @property
+    def sort_field(self) -> KSort:
+        return KSort(f'{self.name}Field')
+
+    @property
+    def sort_method(self) -> KSort:
+        return KSort(f'{self.name}Method')
+
+    @property
+    def klabel(self) -> KLabel:
+        return KLabel(f'contract_{self.name}')
+
+    @property
+    def klabel_method(self) -> KLabel:
+        return KLabel(f'method_{self.name}')
+
+    @property
+    def klabel_field(self) -> KLabel:
+        return KLabel(f'field_{self.name}')
+
+    @property
+    def subsort(self) -> KProduction:
+        return KProduction(KSort('Contract'), [KNonTerminal(self.sort)])
+
+    @property
+    def production(self) -> KProduction:
+        return KProduction(self.sort, [KTerminal(self.name)], klabel=self.klabel)
+
+    @property
+    def macro_bin_runtime(self) -> KRule:
+        return KRule(KRewrite(KEVM.bin_runtime(KApply(self.klabel)), KEVM.parse_bytestack(stringToken(self.bytecode))))
+
+    @property
+    def method_sentences(self) -> List[KSentence]:
+        method_application_production: KSentence = KProduction(KSort('ByteArray'), [KNonTerminal(self.sort), KTerminal('.'), KNonTerminal(self.sort_method)], klabel=self.klabel_method, att=KAtt({'function': ''}))
+        res = [method_application_production]
+        for mprod in [method.production for method in self.methods]:
+            assert isinstance(mprod, KSentence)
+            res.append(mprod)
```

Comment:
This is a recurring pattern, consider extracting a function like this:

```python3
def _check_and_append_sentences(sents: List[KSentence], iterable: Iterable) -> None:
    for sent in iterable:
        assert isinstance(sent, KSentence)
        sents.append(sent)
```

Reply:
Done, it reads much nicer.